common/djangoapps/third_party_auth/saml.py

```diff
@@ -93,3 +95,158 @@ def wrapped_method(*args, **kwargs):
+                'private_key': self.sapsf_private_key,
+            },
+            timeout=10,
+        )
+        assertion.raise_for_status()
+        assertion = assertion.text
+        token = session.post(
+            self.sapsf_token_url,
+            data={
+                'client_id': self.odata_client_id,
+                'company_id': self.odata_company_id,
+                'grant_type': 'urn:ietf:params:oauth:grant-type:saml2-bearer',
+                'assertion': assertion,
+            },
+            timeout=10,
+        )
+        token.raise_for_status()
+        token = token.json()['access_token']
+        session.headers.update({'Authorization': 'Bearer {}'.format(token), 'Accept': 'application/json'})
+        return session
+
+    def get_user_details(self, attributes):
+        """
+        Attempt to get rich user details from the SAP SuccessFactors OData API. If we're missing any
+        of the details we need to do that, fail nicely by returning the details we're able to extract
+        from just the SAML response and log a warning.
+        """
+        details = super(SapSuccessFactorsIdentityProvider, self).get_user_details(attributes)
+        if self.missing_variables():
+            # If there aren't enough details to make the request, log a warning and return the details
+            # from the SAML assertion.
+            return details
+        username = details['username']
+        try:
+            client = self.get_odata_api_client(user_id=username)
+            response = client.get(
+                '{root_url}User(userId=\'{user_id}\')?$select=username,firstName,lastName,defaultFullName,email'.format(
+                    root_url=self.odata_api_root_url,
+                    user_id=username
+                ),
+                timeout=10,
+            )
+            response.raise_for_status()
+            response = response.json()
+        except requests.RequestException:
+            # If there was an HTTP level error, log the error and return the details from the SAML assertion.
+            log.warning(
+                'Unable to retrieve user details with username %s from SAPSuccessFactors with company ID %s.',
+                username,
+                self.odata_company_id,
+            )
+            return details
+        return {
+            'username': response['d']['username'],
+            'first_name': response['d']['firstName'],
+            'last_name': response['d']['lastName'],
+            'fullname': response['d']['defaultFullName'],
+            'email': response['d']['email'],
+        }
+
+
+def get_saml_idp_choices():
+    """
+    Get a list of the available SAMLIdentityProvider subclasses that can be used to process
+    SAML requests, for use in the Django administration form.
+    """
+    return (
+        (STANDARD_SAML_PROVIDER_KEY, 'Standard SAML provider'),
+        ('sap_success_factors', 'SAP SuccessFactors provider'),
```

Comment:
@haikuginger It might be worth introducing a constant for `'sap_success_factors'` as well.